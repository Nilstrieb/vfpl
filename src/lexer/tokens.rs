use crate::error::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub span: Span,
    pub kind: TokenType,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    Please,

    Ident(String),
    // Keywords
    As,
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,
    Initialize,
    Variable,
    With,
    The,
    Value,
    Of,
    Set,
    To,
    From,
    By,
    Take,
    End,
    Check,
    Whether,
    Then,
    Do,
    Otherwise,
    Break,
    Out,
    This,
    Create,
    Function,
    Parameter,
    Parameters,
    That,
    Returns,
    Call,
    No,
    Argument,
    Arguments,
    Go,
    Sleep,
    Absent,
    Null,
    NoValue,
    Undefined,
    True,
    False,
    Does,
    Has,
    Is,

    // literals
    String(String),
    Int(i64),
    Float(f64),

    // symbols
    /// .
    Dot,
    /// (
    ParenOpen,
    /// )
    ParenClose,
    /// ,
    Comma,
}

impl Display for TokenType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenType::Please => f.write_str("keyword `please`"),
            TokenType::Ident(name) => f.write_str(&format!("`{}`", name)),
            TokenType::As => f.write_str("keyword `as`"),
            TokenType::Add => f.write_str("keyword `add`"),
            TokenType::Subtract => f.write_str("keyword `subtract`"),
            TokenType::Multiply => f.write_str("keyword `multiply`"),
            TokenType::Divide => f.write_str("keyword `divide`"),
            TokenType::Modulo => f.write_str("keyword `modulo`"),
            TokenType::Initialize => f.write_str("keyword `initialize`"),
            TokenType::Variable => f.write_str("keyword `variable`"),
            TokenType::With => f.write_str("keyword `with`"),
            TokenType::The => f.write_str("keyword `the`"),
            TokenType::Value => f.write_str("keyword `value`"),
            TokenType::Of => f.write_str("keyword `of`"),
            TokenType::Set => f.write_str("keyword `set`"),
            TokenType::To => f.write_str("keyword `to`"),
            TokenType::From => f.write_str("keyword `from`"),
            TokenType::By => f.write_str("keyword `by`"),
            TokenType::Take => f.write_str("keyword `take`"),
            TokenType::End => f.write_str("keyword `end`"),
            TokenType::Check => f.write_str("keyword `check`"),
            TokenType::Whether => f.write_str("keyword `whether`"),
            TokenType::Then => f.write_str("keyword `then`"),
            TokenType::Do => f.write_str("keyword `do`"),
            TokenType::Otherwise => f.write_str("keyword `otherwise`"),
            TokenType::Break => f.write_str("keyword `break`"),
            TokenType::Out => f.write_str("keyword `out`"),
            TokenType::This => f.write_str("keyword `this`"),
            TokenType::Create => f.write_str("keyword `create`"),
            TokenType::Function => f.write_str("keyword `function`"),
            TokenType::Parameter => f.write_str("keyword `parameter`"),
            TokenType::Parameters => f.write_str("keyword `parameters`"),
            TokenType::That => f.write_str("keyword `that`"),
            TokenType::Returns => f.write_str("keyword `returns`"),
            TokenType::Go => f.write_str("keyword `go`"),
            TokenType::Sleep => f.write_str("keyword `sleep`"),
            TokenType::Absent => f.write_str("keyword `absent`"),
            TokenType::Null => f.write_str("keyword `null`"),
            TokenType::NoValue => f.write_str("keyword `no value` )"),
            TokenType::Undefined => f.write_str("keyword `undefined`"),
            TokenType::True => f.write_str("keyword `true`"),
            TokenType::False => f.write_str("keyword `false`"),
            TokenType::String(value) => f.write_str(&format!("`\"{}\"`", value)),
            TokenType::Int(value) => f.write_str(&format!("`{}`", value)),
            TokenType::Float(value) => f.write_str(&format!("`{}`", value)),
            TokenType::Dot => f.write_str("`.`"),
            TokenType::ParenOpen => f.write_str("`(`"),
            TokenType::ParenClose => f.write_str("`)`"),
            TokenType::Comma => f.write_str("`,`"),
        }
    }
}
