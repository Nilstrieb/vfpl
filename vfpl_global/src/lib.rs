mod global_ctx;
mod session;

pub use global_ctx::GlobalCtx;
pub use session::Session;

/// Only we need to depend on `lasso`, use re-exports for things downstream crates might need
pub use lasso::Spur;
