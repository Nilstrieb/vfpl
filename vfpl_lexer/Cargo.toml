[package]
name = "vfpl_lexer"
version = "0.0.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
vfpl_error = { path = "../vfpl_error" }
vfpl_global = { path = "../vfpl_global" }
fastrand = "1.5.0"
peekmore = { version = "1.0.0", features = ["smallvec"] }
unicode-xid = "0.2.2"

[dev-dependencies]
insta = "1.8.0"