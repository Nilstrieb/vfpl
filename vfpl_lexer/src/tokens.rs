use std::fmt::{Display, Formatter};
use vfpl_error::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub span: Span,
    pub kind: TokenKind,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenKind {
    CondKeyword(CondKeyword),
    Ident(String),
    // Keywords
    Please,
    As,
    Initialize,
    Variable,
    End,
    Check,
    Whether,
    Then,
    Do,
    Otherwise,
    Break,
    This,
    Create,
    Function,
    Call,
    And,
    Absent,
    Null,
    NoValue,
    Undefined,
    True,
    False,
    Not,
    Or,
    Repeat,
    Return,
    While,

    // literals
    String(String),
    Int(i64),
    Float(f64),

    // symbols
    /// .
    Dot,
    /// (
    ParenOpen,
    /// )
    ParenClose,
    /// ,
    Comma,

    /// The last token
    Eof,
}

/// A conditional keyword that can be used as an identifier
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CondKeyword {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    With,
    The,
    Value,
    Of,
    Set,
    To,
    From,
    By,
    Take,
    Out,
    Parameter,
    Parameters,
    That,
    Returns,
    No,
    Argument,
    Arguments,
    Go,
    Sleep,
    Does,
    Has,
    Is,
    Have,
    Greater,
    Less,
    Than,
    Equal,
}

impl CondKeyword {
    fn as_str(&self) -> &'static str {
        match self {
            CondKeyword::Add => "add",
            CondKeyword::Sub => "sub",
            CondKeyword::Mul => "mul",
            CondKeyword::Div => "div",
            CondKeyword::Mod => "mod",
            CondKeyword::With => "with",
            CondKeyword::The => "the",
            CondKeyword::Value => "value",
            CondKeyword::Of => "of",
            CondKeyword::Set => "set",
            CondKeyword::To => "to",
            CondKeyword::From => "from",
            CondKeyword::By => "by",
            CondKeyword::Take => "take",
            CondKeyword::Out => "out",
            CondKeyword::Parameter => "parameter",
            CondKeyword::Parameters => "parameters",
            CondKeyword::That => "that",
            CondKeyword::Returns => "returns",
            CondKeyword::No => "no",
            CondKeyword::Argument => "argument",
            CondKeyword::Arguments => "arguments",
            CondKeyword::Go => "go",
            CondKeyword::Sleep => "sleep",
            CondKeyword::Does => "does",
            CondKeyword::Has => "has",
            CondKeyword::Is => "is",
            CondKeyword::Have => "have",
            CondKeyword::Greater => "greater",
            CondKeyword::Less => "less",
            CondKeyword::Than => "than",
            CondKeyword::Equal => "equal",
        }
    }
}

impl Display for TokenKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenKind::Please => f.write_str("keyword `please`"),
            TokenKind::Ident(name) => f.write_str(&format!("`{}`", name)),
            TokenKind::As => f.write_str("keyword `as`"),
            TokenKind::Initialize => f.write_str("keyword `initialize`"),
            TokenKind::Variable => f.write_str("keyword `variable`"),
            TokenKind::End => f.write_str("keyword `end`"),
            TokenKind::Check => f.write_str("keyword `check`"),
            TokenKind::Whether => f.write_str("keyword `whether`"),
            TokenKind::Then => f.write_str("keyword `then`"),
            TokenKind::Do => f.write_str("keyword `do`"),
            TokenKind::Otherwise => f.write_str("keyword `otherwise`"),
            TokenKind::Break => f.write_str("keyword `break`"),
            TokenKind::This => f.write_str("keyword `this`"),
            TokenKind::Create => f.write_str("keyword `create`"),
            TokenKind::Function => f.write_str("keyword `function`"),
            TokenKind::Absent => f.write_str("keyword `absent`"),
            TokenKind::Null => f.write_str("keyword `null`"),
            TokenKind::NoValue => f.write_str("keyword `novalue` )"),
            TokenKind::Undefined => f.write_str("keyword `undefined`"),
            TokenKind::True => f.write_str("keyword `true`"),
            TokenKind::False => f.write_str("keyword `false`"),
            TokenKind::Not => f.write_str("keyword `not`"),
            TokenKind::Or => f.write_str("keyword `or`"),
            TokenKind::Call => f.write_str("keyword `call`"),
            TokenKind::Repeat => f.write_str("keyword `repeat`"),
            TokenKind::Return => f.write_str("keyword `return`"),
            TokenKind::String(value) => f.write_str(&format!("`\"{}\"`", value)),
            TokenKind::Int(value) => f.write_str(&format!("`{}`", value)),
            TokenKind::Float(value) => f.write_str(&format!("`{}`", value)),
            TokenKind::Dot => f.write_str("`.`"),
            TokenKind::ParenOpen => f.write_str("`(`"),
            TokenKind::ParenClose => f.write_str("`)`"),
            TokenKind::Comma => f.write_str("`,`"),
            TokenKind::And => f.write_str("keyword `and`"),
            TokenKind::While => f.write_str("keyword `while`"),
            TokenKind::Eof => f.write_str("end of file"),
            TokenKind::CondKeyword(cond) => f.write_str(&format!("keyword `{}`", cond.as_str())),
        }
    }
}

impl Token {
    /// Token with a Span of one character
    pub fn new_from_single(kind: TokenKind, idx: usize) -> Self {
        Token {
            kind,
            span: Span::single(idx),
        }
    }

    pub fn new(kind: TokenKind, start: usize, end: usize) -> Self {
        Token {
            kind,
            span: Span::start_end(start, end),
        }
    }

    pub fn new_from_len(kind: TokenKind, start: usize, end: usize) -> Self {
        Token {
            kind,
            span: Span::start_len(start, end),
        }
    }

    pub fn eof() -> Self {
        Token {
            span: Span::eof(),
            kind: TokenKind::Eof,
        }
    }
}
