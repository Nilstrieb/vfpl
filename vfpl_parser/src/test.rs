//// helpers

use super::Parser;
use vfpl_error::Span;
use vfpl_global::{GlobalCtx, Spur};
use vfpl_lexer::tokens::CondKeyword::*;
use vfpl_lexer::tokens::TokenKind::*;
use vfpl_lexer::tokens::{Token, TokenKind};

fn token(kind: TokenKind) -> Token {
    Token {
        span: Span::dummy(),
        kind,
    }
}

#[cfg(test)]
fn ident(str: &str) -> Spur {
    let mut global_context = vfpl_global::GLOBAL_TEST_CTX.lock().unwrap();
    global_context.intern_string(str)
}

/// parses CondKw(The) tokens and appends an EOF token
fn parse<T, F, R>(tokens: T, parse_rule_fn: F) -> R
where
    T: Into<Vec<Token>>,
    F: FnOnce(&mut Parser) -> R,
{
    let mut vec = tokens.into();
    vec.push(Token::eof());
    let mut parser = Parser::new(vec.into_iter(), GlobalCtx::test_ctx());
    parse_rule_fn(&mut parser)
}

//// tests

#[test]
fn it_works() {
    let snap = "test";
    insta::assert_debug_snapshot!(snap);
}

#[test]
fn single_string_literal() {
    let tokens = [String("string".to_string())].map(token);
    let parsed = parse(tokens, Parser::literal);
    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn different_literals() {
    let literals = [
        [String("test".to_string())].map(token),
        [Int(425)].map(token),
        [Float(2465.67)].map(token),
        [Absent].map(token),
        [NoValue].map(token),
        [Null].map(token),
        [Undefined].map(token),
        [True].map(token),
        [False].map(token),
    ];

    let parsed = literals.map(|tokens| parse(tokens, Parser::literal));

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn ident_ty() {
    let tokens = [Ident(ident("Test"))].map(token);
    let parsed = parse(tokens, Parser::ty);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn nullable_ty() {
    let tys = [
        [Absent].map(token),
        [NoValue].map(token),
        [Null].map(token),
        [Undefined].map(token),
    ];
    let parsed = tys.map(|tokens| parse(tokens, Parser::ty));

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn other_tys() {
    let tys = [
        [Ident(ident("Boolean"))].map(token),
        [Ident(ident("Integer"))].map(token),
        [Ident(ident("String"))].map(token),
        [Ident(ident("Float"))].map(token),
    ];
    let parsed = tys.map(|tokens| parse(tokens, Parser::ty));

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn literal_call_expr() {
    let tokens = [String("string".to_string())].map(token);
    let parsed = parse(tokens, Parser::call_expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn literal_expr() {
    let tokens = [String("string".to_string())].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn ident_literal() {
    let tokens = [Ident(ident("uwu"))].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn ident_cond_keyword_literal() {
    let tokens = [CondKw(From)].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn not_equal_expr() {
    let tokens = [
        Absent,
        CondKw(Does),
        Not,
        CondKw(Have),
        CondKw(The),
        CondKw(Value),
        Null,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn equal_expr() {
    let tokens = [Absent, CondKw(Has), CondKw(The), CondKw(Value), Absent].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn greater_than_expr() {
    let tokens = [Absent, CondKw(Is), CondKw(Greater), CondKw(Than), Undefined].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn less_than_expr() {
    let tokens = [Absent, CondKw(Is), CondKw(Less), CondKw(Than), Undefined].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn greater_equal_than_expr() {
    let tokens = [
        Absent,
        CondKw(Is),
        CondKw(Greater),
        Or,
        CondKw(Equal),
        CondKw(Than),
        Undefined,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn less_equal_than_expr() {
    let tokens = [
        Absent,
        CondKw(Is),
        CondKw(Less),
        Or,
        CondKw(Equal),
        CondKw(Than),
        Undefined,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn typed_ident_absent() {
    let tokens = [Ident(ident("name")), As, Absent].map(token);
    let parsed = parse(tokens, Parser::typed_ident);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn init_variable_string() {
    let tokens = [
        Initialize,
        Variable,
        Ident(ident("name")),
        As,
        Ident(ident("string")),
        CondKw(With),
        CondKw(The),
        CondKw(Value),
        CondKw(Of),
        String("Ferris".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::var_init);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn set_variable_string() {
    let tokens = [
        CondKw(Set),
        CondKw(The),
        Variable,
        Ident(ident("name")),
        CondKw(To),
        CondKw(The),
        CondKw(Value),
        CondKw(Of),
        String("Ferris".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::var_set);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn add_number() {
    let tokens = [CondKw(Add), Int(0), CondKw(To), Ident(ident("counter"))].map(token);
    let parsed = parse(tokens, Parser::term);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn subtract_number() {
    let tokens = [CondKw(Sub), Int(1), CondKw(From), Ident(ident("counter"))].map(token);
    let parsed = parse(tokens, Parser::term);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn divide_number() {
    let tokens = [CondKw(Div), Ident(ident("counter")), CondKw(By), Int(2)].map(token);
    let parsed = parse(tokens, Parser::factor);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn multiply_number() {
    let tokens = [CondKw(Mul), Ident(ident("counter")), CondKw(With), Int(3)].map(token);
    let parsed = parse(tokens, Parser::factor);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn mod_number() {
    let tokens = [CondKw(Take), Ident(ident("counter")), CondKw(Mod), Int(4)].map(token);
    let parsed = parse(tokens, Parser::factor);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn nested_add_number() {
    let tokens = [
        CondKw(Add),
        Int(5),
        CondKw(To),
        ParenOpen,
        CondKw(Add),
        Int(6),
        CondKw(To),
        Int(7),
        ParenClose,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::term);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn terminate() {
    let tokens = [CondKw(Go), CondKw(To), CondKw(Sleep)].map(token);
    let parsed = parse(tokens, Parser::terminate);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_no_args() {
    let tokens = [
        Call,
        Ident(ident("run")),
        CondKw(With),
        CondKw(No),
        CondKw(Arguments),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_single_arg() {
    let tokens = [
        Call,
        Ident(ident("print")),
        CondKw(With),
        CondKw(The),
        CondKw(Argument),
        Int(0),
        As,
        Ident(ident("printable")),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_two_args() {
    let tokens = [
        Call,
        Ident(ident("add")),
        CondKw(With),
        CondKw(The),
        CondKw(Arguments),
        Int(2),
        As,
        Ident(ident("a")),
        And,
        Int(3),
        As,
        Ident(ident("b")),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_three_args() {
    let tokens = [
        Call,
        Ident(ident("ternary")),
        CondKw(With),
        CondKw(The),
        CondKw(Arguments),
        True,
        As,
        Ident(ident("cond")),
        Comma,
        Int(0),
        As,
        Ident(ident("then")),
        And,
        Int(3),
        As,
        Ident(ident("else")),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn add_stmt() {
    let tokens = [
        Please,
        CondKw(Add),
        Int(0),
        CondKw(To),
        Ident(ident("A")),
        Dot,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_empty_body_true_cond() {
    let tokens = [Check, Whether, True, Comma, Then, Do, Please, End, Check].map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_empty_body_complex_cond() {
    let tokens = [
        Check,
        Whether,
        True,
        CondKw(Has),
        CondKw(The),
        CondKw(Value),
        Absent,
        Comma,
        Then,
        Do,
        Please,
        End,
        Check,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_single_stmt_body() {
    let tokens = [
        Check,
        Whether,
        True,
        CondKw(Has),
        CondKw(The),
        CondKw(Value),
        Absent,
        Comma,
        Then,
        Do,
        Please,
        CondKw(Go),
        CondKw(To),
        CondKw(Sleep),
        Dot,
        Please,
        End,
        Check,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_multi_stmt_body() {
    let tokens = [
        Check,
        Whether,
        True,
        CondKw(Has),
        CondKw(The),
        CondKw(Value),
        Absent,
        Comma,
        Then,
        Do,
        Please,
        CondKw(Go),
        CondKw(To),
        CondKw(Sleep),
        Dot,
        Please,
        CondKw(Add),
        Int(5),
        CondKw(To),
        Ident(ident("A")),
        Dot,
        Please,
        End,
        Check,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn return_number() {
    let tokens = [Return, Int(9), CondKw(From), CondKw(The), Function].map(token);
    let parsed = parse(tokens, Parser::return_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn return_stmt() {
    let tokens = [
        Please,
        Return,
        Int(9),
        CondKw(From),
        CondKw(The),
        Function,
        Dot,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn r#break() {
    let tokens = [Break, CondKw(Out), CondKw(Of), This, While].map(token);
    let parsed = parse(tokens, Parser::break_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn break_stmt() {
    let tokens = [Please, Break, CondKw(Out), CondKw(Of), This, While, Dot].map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn while_loop_no_body() {
    let tokens = [Repeat, While, True, Do, Please, End, While].map(token);
    let parsed = parse(tokens, Parser::while_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn while_stmt_with_body() {
    let tokens = [
        Please,
        Repeat,
        While,
        Absent,
        CondKw(Has),
        CondKw(The),
        CondKw(Value),
        False,
        Do,
        Please,
        CondKw(Add),
        Int(9),
        CondKw(To),
        Ident(ident("A")),
        Dot,
        Please,
        End,
        While,
        Dot,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn return_ty_absent() {
    let tokens = [CondKw(That), CondKw(Returns), Absent].map(token);
    let parsed = parse(tokens, Parser::fn_return);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_no_params_empty_body() {
    let tokens = [
        Create,
        Function,
        Ident(ident("void")),
        CondKw(With),
        CondKw(No),
        CondKw(Parameters),
        CondKw(That),
        CondKw(Returns),
        Absent,
        Please,
        End,
        Function,
        Ident(ident("void")),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_single_param_emtpy_body() {
    let tokens = [
        Create,
        Function,
        Ident(ident("print")),
        CondKw(With),
        CondKw(The),
        CondKw(Parameter),
        Ident(ident("_")),
        As,
        Absent,
        CondKw(That),
        CondKw(Returns),
        Absent,
        Please,
        End,
        Function,
        Ident(ident("print")),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_two_param_emtpy_body() {
    let tokens = [
        Create,
        Function,
        Ident(ident("add")),
        CondKw(With),
        CondKw(The),
        CondKw(Parameters),
        Ident(ident("_")),
        As,
        Absent,
        And,
        Ident(ident("_hi")),
        As,
        Null,
        CondKw(That),
        CondKw(Returns),
        Absent,
        Please,
        End,
        Function,
        Ident(ident("add")),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_three_param_with_body() {
    let tokens = [
        Create,
        Function,
        Ident(ident("add")),
        CondKw(With),
        CondKw(The),
        CondKw(Parameters),
        Ident(ident("_")),
        As,
        Absent,
        Comma,
        Ident(ident("a")),
        As,
        NoValue,
        And,
        Ident(ident("_hi")),
        As,
        Null,
        CondKw(That),
        CondKw(Returns),
        Absent,
        Please,
        CondKw(Add),
        Int(5),
        CondKw(To),
        Ident(ident("a")),
        Dot,
        Please,
        End,
        Function,
        Ident(ident("add")),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn multiple_body_statements() {
    let tokens = [
        Please,
        CondKw(Add),
        Int(5),
        CondKw(To),
        Int(6),
        Dot,
        Please,
        CondKw(Go),
        CondKw(To),
        CondKw(Sleep),
        Dot,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::body);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn cond_keyword_var_name() {
    let tokens = [
        Initialize,
        Variable,
        CondKw(Add),
        As,
        Ident(ident("string")),
        CondKw(With),
        CondKw(The),
        CondKw(Value),
        CondKw(Of),
        String("Ferris".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::var_init);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn structure_no_field() {
    let tokens = [
        Define,
        Structure,
        Ident(ident("Unit")),
        CondKw(With),
        Please,
        End,
        Define,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::struct_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn structure_single_field() {
    let tokens = [
        Define,
        Structure,
        Ident(ident("NonNullInt")),
        CondKw(With),
        CondKw(The),
        CondKw(Field),
        Ident(ident("inner")),
        As,
        Ident(ident("Integer")),
        Please,
        End,
        Define,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::struct_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn structure_two_fields() {
    let tokens = [
        Define,
        Structure,
        Ident(ident("Person")),
        CondKw(With),
        CondKw(The),
        CondKw(Field),
        Ident(ident("name")),
        As,
        Ident(ident("String")),
        And,
        CondKw(The),
        CondKw(Field),
        Ident(ident("age")),
        As,
        Ident(ident("Integer")),
        Please,
        End,
        Define,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::struct_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn structure_three_fields() {
    let tokens = [
        Define,
        Structure,
        Ident(ident("Person")),
        CondKw(With),
        CondKw(The),
        CondKw(Field),
        Ident(ident("first_name")),
        As,
        Ident(ident("String")),
        Comma,
        CondKw(The),
        CondKw(Field),
        Ident(ident("last_name")),
        As,
        Ident(ident("String")),
        And,
        CondKw(The),
        CondKw(Field),
        Ident(ident("age")),
        As,
        Ident(ident("Integer")),
        Please,
        End,
        Define,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::struct_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn structure_literal_no_fields() {
    let tokens = [
        Ident(ident("Unit")),
        CondKw(With),
        CondKw(No),
        CondKw(Fields),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::struct_literal);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn structure_literal_one_field() {
    let tokens = [
        Ident(ident("NonZeroInt")),
        CondKw(With),
        CondKw(The),
        CondKw(Field),
        Int(3),
        As,
        Ident(ident("inner")),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::struct_literal);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn structure_literal_two_fields() {
    let tokens = [
        Ident(ident("Person")),
        CondKw(With),
        CondKw(The),
        CondKw(Fields),
        Ident(ident("Hugo")),
        As,
        Ident(ident("name")),
        And,
        Int(5),
        As,
        Ident(ident("age")),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::struct_literal);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn structure_literal_three_fields() {
    let tokens = [
        Ident(ident("Person")),
        CondKw(With),
        CondKw(The),
        CondKw(Fields),
        Ident(ident("Hugo")),
        As,
        Ident(ident("first_name")),
        Comma,
        Ident(ident("Boss")),
        As,
        Ident(ident("first_name")),
        And,
        Int(5),
        As,
        Ident(ident("age")),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::struct_literal);

    insta::assert_debug_snapshot!(parsed);
}
