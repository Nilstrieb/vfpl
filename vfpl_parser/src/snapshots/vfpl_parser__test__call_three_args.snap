---
source: vfpl_parser/src/test.rs
expression: parsed

---
Ok(
    Call {
        span: Span {
            start: 0,
            end: 0,
        },
        fn_name: "ternary",
        args: CallArgs {
            span: Span {
                start: 0,
                end: 0,
            },
            args: [
                ValueIdent {
                    span: Span {
                        start: 0,
                        end: 0,
                    },
                    expr: Literal(
                        Literal {
                            span: Span {
                                start: 0,
                                end: 0,
                            },
                            kind: True,
                        },
                    ),
                    name: "cond",
                },
                ValueIdent {
                    span: Span {
                        start: 0,
                        end: 0,
                    },
                    expr: Literal(
                        Literal {
                            span: Span {
                                start: 0,
                                end: 0,
                            },
                            kind: Int(
                                0,
                            ),
                        },
                    ),
                    name: "then",
                },
                ValueIdent {
                    span: Span {
                        start: 0,
                        end: 0,
                    },
                    expr: Literal(
                        Literal {
                            span: Span {
                                start: 0,
                                end: 0,
                            },
                            kind: Int(
                                3,
                            ),
                        },
                    ),
                    name: "else",
                },
            ],
        },
    },
)
